### ISSUE TITLE
Exposing Functions on Web Components

### ISSUE THREAD

**Issue by devpaul** [dojo/framework#253]
I would like to create a video player using Dojo that encapsulates a `<video>` tag with some additional functionality; e.g. an ad overlay, title, description, and credit for the author. I'd like to distribute this widget as a web component that has a set of APIs for interacting with the widget: `load()`, `play()`, `skip30()`, `back30()`, `pause()`, `stop()`. What is the best way for defining a set of behaviors on a Dojo web component?

**code**

example usage of web component to skip ahead one minute and play:

```js
const player = document.querySelector('my-video-component');
player.pause();
player.skip30();
player.skip30();
player.play();
```

**Comment by matt-gadd** [comment]
We don't support this and we have no plans to as part of the widget system or cli-build-widget. We're fixing up some of the documentation currently for dojo.io, and when we get to custom elements we should definitely document this as a known limitation/limitation by design. 

To clarify why, the goal of the custom element target for widgets is to take a standard dojo widget and transparently turn it into a custom element without any user intervention or specific custom element code. We already mostly do this other than the fact the user has to specify the tag/properties/attributes/events in a decorator, but shortly when https://github.com/dojo/webpack-contrib/issues/43 has landed, the decorator will be redundant and a user can then take any dojo widget and target as a custom element.

Given the scenario (which is i'd say uncommon for most things other than media elements, and doesn't fit well as a reactive dojo widget in the first place), you could still expose a programatic api just not off the element directly:

```ts
const player = document.querySelector('my-video-component');
player.controls = ({ play, pause, skip30 }) => {
  pause();
  skip30();
  skip30();
  play();
}
```

I don't think this is massively unreasonable, but still wouldn't recommend it due to the first statement - that it introduces imperative/custom element specific code into the widget (although I guess the controlling code could at least be a wrapping widget).

The other option in these bespoke scenario's would be to write a wrapper around the element. You could do this by leveraging `@dojo/framework/widget-core/registerCustomElement`'s exported `create()` function, which does not register the element, that should then allow you to wrap the element with whatever api's and register it yourself but I think you'll then be on your own in terms of building the element etc. It seems this would be trivial enough for you to use the existing build by ejecting cli-build-widget and modifying the template here https://github.com/dojo/cli-build-widget/blob/master/src/template/custom-element.js.

To summarise, this isn't one of the design goals of the widget custom element target.

**Comment by devpaul** [comment]
This seems like a great approach. `registerCustomElement()` used to encapsulate the custom element, but now `create()` makes it accessible so it can be further extended.

So it should be possible to write something like this in a `main.ts` and compile it like a regular dojo app.

```typescript
class MyVideoPlayer extends create(descriptor, WidgetConstructor) {
    play() { }
    skip30() { }
    pause() { }
}

global.customElements.define('my-video-component', MyVideoPlayer);
```

There is still the challenge of converting an imperative method call into a declarative property. I think this could be done using a signal pattern that works in tandem w/ a `meta` on the widget. I've been experimenting with [passing signals](https://github.com/devpaul/dojo-clippy/blob/agent-refactor/src/agents/agent/Animator.ts#L13) and then [handling a signal](https://github.com/devpaul/dojo-clippy/blob/agent-refactor/src/agents/agent/Animator.ts#L31) via a decorated method. It might be useful for anyone facing this challenge.

Here's a small example using a signal and a meta to play audio (`Signal` and `@signal` are not part of the dojo library, their implementation is available from the links above).

```typescript
export interface PlayerProperties extends WidgetProperties {
	play?: Signal<string>;
}

export class Animator<P extends PlayerProperties = PlayerProperties > extends WidgetBase<P> {
	@signal('play')
	onPlay(action: string) {
		if (this.isLoaded) {
			this.meta(AudioMeta).play('audio');
		}
	}

	protected render(): DNode | DNode[] {
		const { src } = this.properties;

		return v('audio', {
			src,
			key: 'audio'
		});
	}
}
```

I think your answer @matt-gadd gives a really good starting point for Dojo, a reactive architecture, to provide imparative functions. I'm going to close this issue. Thanks 👍
